server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -725,12 +836,107 @@ private void handleFinalizationFailure(Exception e, SnapshotsInProgress.Entry en
+     * Run the next queued up repository operation for the given repository name.
+     *
+     * @param newGeneration current repository generation for given repository
+     * @param repository    repository name
+     */
+    private void runNextQueuedOperation(long newGeneration, String repository) {
+        synchronized (currentlyFinalizing) {
+            assert currentlyFinalizing.contains(repository);
+            final Deque<SnapshotFinalization> outstandingForRepo = snapshotsToFinalize.get(repository);
+            final SnapshotFinalization nextFinalization;
+            if (outstandingForRepo == null) {
+                nextFinalization = null;
+            } else {
+                nextFinalization = outstandingForRepo.pollFirst();
+                if (outstandingForRepo.isEmpty()) {
+                    snapshotsToFinalize.remove(repository);
+                }
+            }
+            if (nextFinalization == null) {
+                final boolean removed = currentlyFinalizing.remove(repository);
+                assert removed;
+                runReadyDeletions();
+            } else {
+                logger.trace("Moving on to finalizing next snapshot [{}]", nextFinalization);
+                finalizeSnapshotEntry(nextFinalization.entry, nextFinalization.metadata, newGeneration);
+            }
         }
     }
 
+    /**
+     * Runs a cluster state update that checks whether we have outstanding snapshot deletions that can be executed and executes them.
+     *
+     * TODO: optimize this to execute in a single CS update together with finalizing the latest snapshot
+     */
+    private void runReadyDeletions() {
+        clusterService.submitStateUpdateTask("Run ready deletions", new ClusterStateUpdateTask() {
+
+            private List<SnapshotDeletionsInProgress.Entry> deletionsToRun;
+
+            @Override
+            public ClusterState execute(ClusterState currentState) {
+                final Tuple<ClusterState, List<SnapshotDeletionsInProgress.Entry>> res = readyDeletions(currentState);
+                assert res.v1() == currentState : "Deletes should have been set to ready by finished snapshot deletes and finalizations";
+                deletionsToRun = res.v2();
+                return res.v1();
+            }
+
+            @Override
+            public void onFailure(String source, Exception e) {
+                logger.warn("Failed to run ready delete operations", e);
+            }
+
+            @Override
+            public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
+                for (SnapshotDeletionsInProgress.Entry entry : deletionsToRun) {
+                    deleteSnapshotsFromRepository(entry, entry.repositoryStateId(), newState.nodes().getMinNodeVersion());
+                }
+            }
+        });
+    }
+
+    private static Tuple<ClusterState, List<SnapshotDeletionsInProgress.Entry>> readyDeletions(ClusterState currentState) {
```

Comment:
add javadocs here to explain what this method does?

Reply:
Done :)